services-custom/s3-transfer-manager/src/main/java/software/amazon/awssdk/transfer/s3/internal/UploadDirectoryManager.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.transfer.s3.internal;
+
+
+import java.io.IOException;
+import java.nio.file.FileVisitOption;
+import java.nio.file.Files;
+import java.nio.file.LinkOption;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.transfer.s3.CompletedUpload;
+import software.amazon.awssdk.transfer.s3.CompletedUploadDirectory;
+import software.amazon.awssdk.transfer.s3.FailedUpload;
+import software.amazon.awssdk.transfer.s3.S3TransferManager;
+import software.amazon.awssdk.transfer.s3.Upload;
+import software.amazon.awssdk.transfer.s3.UploadDirectory;
+import software.amazon.awssdk.transfer.s3.UploadDirectoryRequest;
+import software.amazon.awssdk.transfer.s3.UploadRequest;
+import software.amazon.awssdk.utils.CompletableFutureUtils;
+import software.amazon.awssdk.utils.Logger;
+import software.amazon.awssdk.utils.StringUtils;
+
+@SdkInternalApi
+public class UploadDirectoryManager {
+    private static final Logger log = Logger.loggerFor(S3TransferManager.class);
+
+    private final TransferConfiguration transferConfiguration;
+    private final Function<UploadRequest, Upload> uploadFunction;
+
+    public UploadDirectoryManager(TransferConfiguration transferConfiguration,
+                                  Function<UploadRequest, Upload> uploadFunction) {
+
+        this.transferConfiguration = transferConfiguration;
+        this.uploadFunction = uploadFunction;
+    }
+
+    public UploadDirectory uploadDirectory(UploadDirectoryRequest uploadDirectoryRequest) {
+        CompletableFuture<CompletedUploadDirectory> returnFuture = new CompletableFuture<>();
+        Path directory = uploadDirectoryRequest.sourceDirectory();
+
+        List<CompletedUpload> uploads = new ArrayList<>();
+        List<FailedUpload> failedUploads = new ArrayList<>();
+        List<CompletableFuture<CompletedUpload>> uploadFutures;
```

Comment:
If we get a feature request in the future where someone says "I'm uploading millions of files and I'm running out of memory", is that something we can accommodate? E.g. by adding a "don't record file-level results" flag? I don't see why not, but you're closer to this than me.